src/document/printer.js

```diff
@@ -142,27 +142,41 @@ function generateInd(ind, newPart, options) {
   }
 }
 
-// TODO: Skip cursor placeholder
 // Trim `Tab(U+0009)` and `Space(U+0020)` at the end of line
 function trim(out) {
   let trimCount = 0;
+  let cursorCount = 0;
+  let outIndex = out.length;
 
-  while (out.length > 0) {
-    const last = getLast(out);
-    if (typeof last !== "string") {
-      break;
+  outer: while (outIndex--) {
+    const last = out[outIndex];
+
+    if (last === CURSOR_PLACEHOLDER) {
+      cursorCount++;
+      continue;
     }
 
-    const trimmed = last.replace(/[\t ]*$/, "");
-    trimCount += last.length - trimmed.length;
+    if (process.env.NODE_ENV !== "production" && typeof last !== "string") {
+      throw new Error(`Unexpected value in trim: '${typeof last}'`);
+    }
 
-    if (trimmed !== "") {
-      out[out.length - 1] = trimmed;
-      break;
+    for (let charIndex = last.length - 1; charIndex >= 0; charIndex--) {
```

Comment:
How do you think add a function to "get trailing space/tab length" so we don't need use label to break.

Reply:
If avoiding the label is the only benefit, I'm not convinced.